keps/sig-cli/4764-kubectl-watchlist/README.md

```diff
@@ -0,0 +1,757 @@
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Feature gate][feature gate] lifecycle
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
+or by redefining what graduation means.
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Additional tests are in Testgrid and linked in KEP
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+### Upgrade / Downgrade Strategy
+
+This feature will not be enabled by default in kubectl until the dependent featuregate in the apiserver 
+are enabled by default.
+
+Upgrading requires that the Kubelet's environment variable be configured to `true`, and the apiserver must
+enable `WatchList` and `ConsistentListFromCache`, as well as disable `WatchFromStorageWithoutResourceVersion`. 
+
+Downgrade requires kubectl environment variable be configured to `false`.
+
+### Version Skew Strategy
+
+This feature will not be enabled by default in kubectl until the dependent featuregate in the apiserver 
+are enabled by default. However, if users actively enable it while using an older version of the apiserver,
+the following scenarios may occur:
+
+#### `WatchList` is not supported or disabled
+
+When the API is on an older version or the dependent features have not yet been enabled, the `sendInitialEvent` 
```

Comment:
AIUI the API verb is **watch** for a watch list fetch; maybe I'm wrong though? Let's make it clear which we expect.
```suggestion
When the API server is on an older version or the dependent features have not yet been enabled,
the `sendInitialEvent` parameter in the **watch**request will not be correctly processed.
An API server prior to version 1.27 will ignore that query parameter and treat the request as a standard
**watch** request, which may lead to unexpected behavior in kubectl unless correctly handled.
In kube-apiserver version 1.27 and later, where the `WatchList` feature gate is disabled, the API server will respond with an error as expected; kubectl can then fallback to the original **list** verb and associated
client side logic.
```